design/service-binding.md

```diff
@@ -0,0 +1,356 @@
+
+``` yaml
+apiVersion: operators.coreos/com:v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  name: strimzi-cluster-operator.v0.16.2
+  namespace: placeholder
+  annotations:
+    containerImage: docker.io/strimzi/operator:0.16.2
+spec:
+  customresourcedefinitions:
+    owned:
+    - description: Represents a Kafka cluster
+      displayName: Kafka
+      kind: Kafka
+      name: kafkas.kafka.strimzi.io
+      version: v1beta1
+      statusDescriptors:
+      - description: The addresses of the bootstrap servers for Kafka clients
+        displayName: Bootstrap servers
+        path: status.listeners.{{servicebinding:listener}}.bootstrap
+        x-descriptors:
+        - 'urn:alm:descriptor:servicebinding:secret:endpoint'
+```
+
+Note the highly irregular `{{servicebinding:listener}}` syntax which I've used to mean that the Service Binding Operator has a variable `listener` provided as part of a `ServiceBindingRequest`, and it can then use this to build the full status path, such as `status.listeners.plain.bootstrap`.
+
+``` yaml
+apiVersion: service.binding/v1alpha1
+kind: ServiceBindingRequest
+metadata:
+  name: my-binding
+  data:
+    listener: plain
+spec:
+  ...
+```
+
+CSV has a limitation here in its descriptor path capability, so **OLM would also need enhancing**. Let's try another way.
+
+## Strimzi enhancement - Specify listener in KafkaUser
+
+When authentication is being used, the user is likely to be used a `KafkaUser` custom resource. This makes quite a good fit for service binding because it is essentially Strimzi's way of expressing a binding.
+
+``` yaml
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaUser
+metadata:
+  name: my-user
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  authentication:
+    type: scram-sha-512
+  authorization:
+    type: simple
+    acls:
+      - resource:
+          type: topic
+          name: my-topic
+          patternType: literal
+        operation: Read
+status:
+  username: my-user-name
+  secret: my-user
+```
+
+The secret looks like this:
+
+``` yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: my-user
+  labels:
+    strimzi.io/kind: KafkaUser
+    strimzi.io/cluster: my-cluster
+type: Opaque
+data:
+  password: Z2VuZXJhdGVkcGFzc3dvcmQ=
+```
+
+One potential way to make this easier, would be to name the target listener and include the bootstrap information in the `KafkaUser` CR like this:
+
+``` yaml
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaUser
+metadata:
+  name: my-user
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  listener: external
+  authentication:
+    type: scram-sha-512
+  authorization:
+    type: simple
+    acls:
+      - resource:
+          type: topic
+          name: my-topic
+          patternType: literal
+        operation: Read
+status:
+  username: my-user-name
+  secret: my-user
+  listener:
+    bootstrap: myhost1.example.com:9092,myhost2.example.com:9092
```

Comment:
This would need to be an array and not a single field since this is currently 0 to 3 listeners, in the future possibly even more. 

This is good idea overall, but might be more complicated than it looks. The User Operator works independently on the Kafka cluster configuration. So it doen't really know whether there are any listeners which use given authentication mechanism or not, eventually which listeners they are and what is their address.

Reply:
Somehow, I'd like a user to be able to create a CR that gets them credentials to Strimzi that also gives a clear, unambiguous way of obtaining the endpoint. I'd like the `spec` to provide the desired listener name so that the SBO doesn't need to wade through an array of listeners.

Reply:
The problem is that the KafkaUser does not belong to a single listener. An it would be probably quite hard to make it so. So IMHO there is no way how to deal about this without a list

Reply:
Ah, I see. I was thinking of the listener name in the KafkaUser as a preference and not a restriction. Kind of "I want a KafkaUser and I'm going to use it for this purpose". I didn't expect to introduce an extra security check to restrict use of credentials to a specific user.

Reply:
I think the problem is that you do not know what the user prefers. You can probably say that the TLS one is prefered over PLAINTEXT. But you cannot say whether the user will be used witnin the Kube cluster or from outside etc. It could be also misleading because ultimately it might suggest that the preffered is the supported one.

I think it would still be a nice feature to have the bootstrap servers there. I think it is real useful. The only thing to keep in mind is that this is non-trivial work really since we need to pass the information to the User OPerator in the first place.

Reply:
This comment has been reflected in the latest version of the document. Multiple listeners are supported.